### ISSUE TITLE
[Feature Request] `nuitka` support

### ISSUE THREAD

**Issue by shenwpo** [beartype/beartype#197]
### ***Beartype0.11 is incompatible with nuitka***

*env*
```
Ubuntu 20.04
python 3.8
nuitka 1.2.7
beartype 0.11
```
*Code*:
```
import typing

from beartype.door import TypeHint


hint = typing.Union[int, float]
type_hints = TypeHint(hint)
for t in type_hints:
    print(t)
```
*build*
```
python3 -m nuitka --onefile main.py
```
*error*
```
./main.bin 
Traceback (most recent call last):
  File "/tmp/.mount_main.bGqEond/main.py", line 3, in <module>
    from beartype.door import TypeHint
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/tmp/.mount_main.bGqEond/beartype/__init__.py", line 57, in <module beartype>
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/tmp/.mount_main.bGqEond/beartype/_decor/decormain.py", line 24, in <module beartype._decor.decormain>
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/tmp/.mount_main.bGqEond/beartype/_data/datatyping.py", line 26, in <module beartype._data.datatyping>
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "/tmp/.mount_main.bGqEond/beartype/_cave/_cavefast.py", line 257, in <module beartype._cave._cavefast>
  File "/tmp/.mount_main.bGqEond/beartype/_cave/_cavefast.py", line 250, in _closure_varcell_factory
TypeError: 'NoneType' object is not subscriptable
```

### nuitka related [issues:](https://github.com/Nuitka/Nuitka/issues/1674)

[incompatible code](https://github.com/beartype/beartype/blob/main/beartype/_cave/_cavefast.py#L233)
```
# ....................{ TYPES ~ call : closure             }....................
def _closure_varcell_factory():
    '''
    Arbitrary function returning a closure-specific cell variable exposed by an
    arbitrary closure isolated to this function.
    '''

    # Arbitrary outer local variable.
    cell_variable = 42

    def closure():
        '''
        Arbitrary closure exposing a closure-specific cell variable.
        '''

        nonlocal cell_variable

    return closure.__closure__[0]  # pyright: ignore[reportOptionalSubscript]


# Although Python >= 3.7 now exposes an explicit closure cell variable type via
# the standard "types.CellType" object, this is of no benefit to older versions
# of Python. Ergo, the type of an arbitrary method wrapper guaranteed to
# *ALWAYS* exist is obtained instead.
ClosureVarCellType = type(_closure_varcell_factory())
```

**Comment by leycec** [comment]
**Gah!** `nuitka` incompatibilities are *definitely* concerning. On the one hand, `nuitka` choking on core Python types is also *definitely* concerning. On the other hand, the @beartype code in question that `nuitka` is complaining about is entirely incidental and can be readily disposed of by just replacing the above logic with:

```python
ClosureVarCellType = types.CellType
```

On the [gripping hand](http://www.catb.org/~esr/jargon/html/O/on-the-gripping-hand.html), the standard library logic defining the `types.CellType` attribute is *basically* a one-for-one verbatim copy of the @beartype logic defining the `ClosureVarCellType` attribute above. In other words, `nuitka` will still probably fail on `types.CellType`. Moreover, given that `nuitka` can't even get past the early-time `beartype._cave` subpackage without choking on its own internal sputum, it kinda seems unlikely that any amount of herculean effort on @beartype's part will allow `nuitka` to compile @beartype.

That is super-disappointing. I know! Our team was really looking forward to dropping Cython in favour of `nuitka`, because Cython is just **so much work.** But @beartype might be *too* magical for `nuitka`. @beartype does many mystical and dark things at runtime – like dynamically defining classes and functions on-the-fly as well as the obligatory metaclass madness that even I shudder spasmodically to maintain.

In short, @beartype is a torture test for Python compilers. `nuitka` just doesn't appear to be complete enough to support the pain that @beartype brings. Nonetheless, let's add `nuitka` support to @beartype's long-term roadmap. It'd be so noice if `nuitka` compiled @beartype out-of-the-box, right?

Thanks so much for the detailed issue, @shenwpo! Lovin' that flaming-letter-e avatar, too! So. Metal. :man_singer:

**Comment by leycec** [comment]
**Celebrations!** I spoke gibberish too soon. After trivially replacing the problematic logic you noted with just `ClosureVarCellType = types.CellType`, `nuitka` now seems to at least superficially support the `beartype.door` API:

```bash
$ python3 -m nuitka --onefile nuitka.py
TypeHint(<class 'int'>)
TypeHint(<class 'float'>)
```

I then expanded your minimal-length example (MLE) to this:

```python
from beartype import beartype
from beartype.door import TypeHint
from beartype.typing import Tuple, Union

@beartype
def make_type_hints() -> Tuple[TypeHint, ...]:
    hint = Union[int, float]
    type_hints = TypeHint(hint)
    return tuple(t for t in type_hints)

for type_hint in make_type_hints():
    print(type_hint)
```

Surprisingly, `nuitka` still quietly behaves itself in the corner:

```bash
$ python3 -m nuitka --onefile nuitka.py
TypeHint(<class 'int'>)
TypeHint(<class 'float'>)
```

:open_mouth:

**Comment by leycec** [comment]
**Possibly resolved by d57a29b.** If you'd like to test @beartype against `nuitka` a bit more for us all, that would be wonderful! Just:

```bash
python3 -m pip install git+https://github.com/beartype/beartype.git@d57a29bdcbfc081cb97bd51789274941d9c60cd7
```

In the meanwhile, I'll continue iterating on this feature request. Specifically, I'll add an integration test exercising @beartype against `nuitka`. Thanks again for the awesome suggestion and example, @shenwpo! :full_moon_with_face:

**Comment by shenwpo** [comment]
@leycec, thank you for your timely reply! Using `ClosureVarCellType = types.CellType` seems to work normally, but `types.CellType` is added in Python [3.8.](https://docs.python.org/3.11/library/types.html#types.CellType)

**Comment by leycec** [comment]
_ohnoes._ I mistakenly assumed that `types.CellType` was only added in Python 3.7. I'll need to carefully add back the prior logic for Python 3.7, which means that @beartype will only be compatible with `nuitka` under Python ≥ 3.8. I'm sorta fine with that, because Python 3.7 is *almost* officially dead. I'll repair that quick. :factory_worker: 

Oh – and it's *super* exciting that you'll be using `beartype.door`, too. That makes you our first official `TypeHint` customer! Please let me know in another issue if you hit any significant issues there. It's mostly well-tested, but early adopters may stumble into unforeseen API pitfalls. :hole:

**Comment by leycec** [comment]
**Resolved by 475dda6...** I *think.* I've added a new `test_nuitka()` integration test to our `pytest` test suite exercising that `nuitka` successfully compiles the example script above. This test is so painfully slow, though, that I've had to temporarily disable it – defeating the whole point of writing a test! Seriously, though. That test currently takes like twenty minutes on my ailing 2010-era AMD Athlon(tm) II X2 240. :face_exhaling: 

Let me know if you hit any additional snags with @beartype under `nuitka`. Until then, thanks again for this fascinating detour into the musty corridors of Python transpilation. You're great, @shenwpo!
